### PR TITLE
Create a K8s app to deploy cncf/devstats for Kubeflow.

### PR DESCRIPTION
**Pull request by jlewi**
* Create a Docker image with the cncf/devstats binaries
* Use a ConfigMap to store the project config
* Add stateful sets for influxdb and postgre
* Add a stateful set to deploy devstats container so that we can do
  kubectl exec to run commands.

/assign @gaocegege 
/cc @lukaszgryglicki 

Related to #34

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/community/91)

### CODE REVIEW

File: devstats/README.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+./runq util_sql/top_unknowns.sql {{ago}} '1 month' {{lim}} 10
+```
+
+### Current status
+
+Running `devstats` in the devstats-cli container to sync the data currently produces the following error.
+
+```
+Error(time=2018-04-18 18:39:01.274140856 +0000 UTC):
+Error: 'pq: database "devstats" does not exist'
```
Comment (lukaszgryglicki) [inline devstats/README.md]: `devstats` database is a separate postgres database that only holds logs (`gha_logs` table).
Its setup is described in INSTALL_*.md files.

File: devstats/devstats_cli.yaml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        # Host will follow statefulset conventions
+        - name: PG_HOST
+          value: postgre-0.ghadb.devstats.svc.cluster.local
+        #  Keep in sync with the values on postgress container
+        #  Also some values presumably can't be changed without recreating the DB.
+        - name: PG_DB
+          value: gha
+        - name: PG_PASS
+          value: password
+
```
Comment (lukaszgryglicki) [inline devstats/devstats_cli.yaml]: You probably need to refer logs database here somehow.
You need two Postgres databases minmally:
- Your project's database - can have any name
- Logs database named `devstats` containing single table `gha_logs`, see INSTALL_*.md files.

File: devstats/ghadb_postgres.yaml [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        image: postgres:10.1-alpine
+        env:
+        # For a list of POSTGRES environment variables see
+        # https://hub.docker.com/_/postgres/
+        - name: POSTGRES_USER
+          value: gha_admin
+        - name: POSTGRES_PASSWORD
+          value: password
+        - name: POSTGRES_DB
+          value: gha
```
Comment (lukaszgryglicki) [inline devstats/ghadb_postgres.yaml]: Probably logs database also needed here (or maybe logs database will be in a different pod/yaml/deployment)?

File: devstats/projects.yaml [hunk 4]
```diff
@@ -0,0 +1,14 @@
+projects:
+  kubeflow:
+    order: 1
+    command_line:
+      - kubeflow
+    # TODO(jlewi): Move the start date back in time once we are done doing some initial testing
+    start_date: 2018-04-16T00:00:00Z
+    # Can we get rid of join date since we aren't part of CNCF yet?
+    join_date: 2018-04-16T00:00:00Z
```
Comment (lukaszgryglicki) [inline devstats/projects.yaml]: You can set it to `null` just like opencontainers project in CNCF repo:
See: https://github.com/cncf/devstats/blob/master/projects.yaml#L241

File: devstats/Dockerfile.devstats [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+RUN apt-get update -y && \
+	apt-get install -y apt-transport-https
+
+RUN echo "deb https://repos.influxdata.com/debian jessie stable" | tee -a /etc/apt/sources.list && \
+	apt-get update && \
+	apt-get install influxdb -y
+
+# Add /go/src/devstats to the path so that all the binaries will be on the path.
+# This is needed by the devstats binary.
+ENV PATH $PATH:/go/src/devstats
```
[the remark this answers is not in the document]
Comment (jlewi) [reply devstats/Dockerfile.devstats]: No problem; wasn't expecting you to. Thanks for the comments.

File: devstats/Dockerfile.devstats [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+	go get github.com/jgautheron/usedexports && \
+	go get github.com/kisielk/errcheck && \
+	go get github.com/influxdata/influxdb/client/v2 && \
+	go get github.com/lib/pq && \
+	go get golang.org/x/text/transform && \
+	go get golang.org/x/text/unicode/norm && \
+	go get gopkg.in/yaml.v2 && \
+	go get github.com/google/go-github/github && \
+	go get golang.org/x/oauth2 && \
+	go get github.com/mattn/go-sqlite3
```
Comment (gaocegege) [inline devstats/Dockerfile.devstats]: BTW, @lukaszgryglicki Could we add vendor in devstats side in the future thus we could avoid `go get` in the docker image.

File: devstats/Dockerfile.devstats [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+RUN cd ${GOPATH}/src && \
+	git clone https://github.com/cncf/devstats.git devstats
+
+RUN cd ${GOPATH}/src/devstats && \
+	git checkout v0.4.0 && \
```
[the remark this answers is not in the document]
Comment (jlewi) [reply devstats/Dockerfile.devstats]: Done.

File: devstats/Makefile [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+	docker tag $(IMG):$(TAG) $(IMG):latest
+	@echo Built $(IMG):$(TAG) and tagged with latest
+
+# Start the postgre container locally for debugging
+# 
+# TODO(jlewi): Not sure this is all that useful. Ideally we should just make
+# this a ksonnet app and run locally on minikube.
+run_postgre:
+	-docker rm -f postgres
+	docker run -d --name=postgres -p 5432:5432 -e POSTGRES_USER=gha_admin -e POSTGRES_PASSWORD=password -e POSTGRES_DB=gha  postgres:10.1-alpine
```
[the remark this answers is not in the document]
Comment (jlewi) [reply devstats/Makefile]: Thanks.

File: devstats/README.md (hunk 1 above)
Comment (jlewi) [reply devstats/README.md]: Done; thank you.

File: devstats/devstats_cli.yaml (hunk 2 above)
Comment (jlewi) [reply devstats/devstats_cli.yaml]: I have two databases both stored in a single instance of postgre.

**Comment by jlewi** [comment]
Thanks PTAL I made significant changes.

File: devstats/README.md (hunk 1 above)
Comment (lukaszgryglicki) [reply devstats/README.md]: BTW I'm going to update auto deploy script - to check if that DB exists. If not (it is shared for all projects) assume that this is a "from zero" setup - create database, admin, ro_user, team role etc.

File: devstats/Dockerfile.devstats [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+# Add postgres to the sudoers group because some of the devstats scripts require it
+RUN adduser postgres sudo
+
+RUN ln -sf /go/src/devstats/util_sql /etc/gha2db/util_sql && \
+	ln -sf /go/src/devstats/metrics /etc/gha2db/metrics && \
+	# /etc/projects-volume will be volume mounted into the pod.
+	ln -sf /etc/projects-volume/projects.yaml /etc/gha2db/projects.yaml
+
+# TODO(jlewi): Per the instructions for devstats we should increase the number of default connections for postgres
```
Comment (gaocegege) [inline devstats/Dockerfile.devstats]: I think we should set the postgre-docker-entrypoint.sh as entrypoint.

File: devstats/Dockerfile.devstats [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+	&& ln -sv ../postgresql.conf.sample "/usr/share/postgresql/$PG_MAJOR/" \
+	&& sed -ri "s!^#?(listen_addresses)\s*=\s*\S+.*!\1 = '*'!" /usr/share/postgresql/postgresql.conf.sample
+
+RUN mkdir -p /var/run/postgresql && chown -R postgres:postgres /var/run/postgresql && chmod 2777 /var/run/postgresql
+
+ENV PATH $PATH:/usr/lib/postgresql/$PG_MAJOR/bin
+
+RUN mkdir -p /home/
+COPY postgre-docker-entrypoint.sh /usr/local/bin/
+RUN chmod a+x /usr/local/bin/postgre-docker-entrypoint.sh
```
Comment (gaocegege) [inline devstats/Dockerfile.devstats]: Why do we set the permission in docker image? Could we set it outside and copy it into the image or just set `bash /usr/local/bin/postgre-docker-entrypoint.sh` as entrypoint?

File: devstats/Dockerfile.devstats (hunk 6 above)
Comment (gaocegege) [reply devstats/Dockerfile.devstats]: @jlewi https://github.com/cncf/devstats/pull/99 devstas has vendor now and maybe we could not install so many deps in the docker image.

File: devstats/Dockerfile.devstats (hunk 9 above)
Comment (jlewi) [reply devstats/Dockerfile.devstats]: The docker image is used to run other things; like the CLI stateful set. Which is why I didn't want to set the entrypoint here and instead do it in the pod spec.

File: devstats/Dockerfile.devstats (hunk 10 above)
Comment (jlewi) [reply devstats/Dockerfile.devstats]: I don't think permissions are preserved when you copy the file. Usually I have to do chmod a+x to make scripts executable inside the container.

File: devstats/Dockerfile.devstats (hunk 9 above)
Comment (gaocegege) [reply devstats/Dockerfile.devstats]: SGTM

File: devstats/Dockerfile.devstats (hunk 10 above)
Comment (gaocegege) [reply devstats/Dockerfile.devstats]: SGTM

Comment (jlewi) [reply devstats/Dockerfile.devstats]: Thanks.

**Comment by jlewi** [comment]
@gaocegege LGTM?

**Pull request merged**